apps/frontend/src/layout/PrivateRoute.tsx

```diff
@@ -1,11 +1,243 @@
     const { token } = useAuthToken();
+    const { organizationSlug, projectId } = useParams();
+    const { viewer, createSession, updateSession } = useSession();
+    const navigate = useNavigate();
+    const [fetchViewer, { loading }] = useViewerLazyQuery();
 
-    return <>{!token ? <Navigate to="/" /> : children}</>;
+    useEffect(() => {
+        let user: SessionViewer | null = viewer || null;
+        let viewerIsLatest = false;
+        let shouldRedirect = false;
+        let org: OrganizationCountableEdge | null = null;
+
+        const validateUrl = async () => {
+            if (!token) {
+                return;
+            }
+            logDebug("viewer: ", user);
+            if (!user) {
+                logDebug(
+                    "No viewer, checking local storage...",
+                    "PrivateRoute",
+                );
+                const getCachedViewer = localStorage.getItem("viewer");
+                if (getCachedViewer) {
+                    user = JSON.parse(getCachedViewer) as SessionViewer;
+                    createSession(user);
+                } else {
+                    logDebug("No viewer in local storage, fetching viewer...");
+                    await fetchViewer({
+                        onCompleted: ({ viewer }) => {
+                            logDebug("Gotten viewer, updating session...");
+                            user = omitTypename(viewer);
+                            createSession(user);
+                            viewerIsLatest = true;
+                        },
+                    });
+                }
+            }
+
+            if (!user) {
+                logDebug("No user, redirecting to login...");
+                navigate("/");
+                return;
+            }
+
+            logDebug("Viewer check passed, checking org...");
+
+            // Org checks
+            if (organizationSlug !== user.organization?.slug) {
+                logDebug(
+                    "Requested organization is not the current organization,\nChecking if user belongs to this organization",
+                );
+                org =
+                    user.organizations?.edges.find((edge) => {
+                        return edge.node.slug === organizationSlug;
+                    }) ?? null;
+
+                if (org) {
+                    user.organization = {
+                        id: org.node.id,
+                        memberCount: org.node.memberCount,
+                        name: org.node.name,
+                        slug: org.node.slug,
+                    };
+                    updateSession(user);
+                }
+
+                if (!org && viewerIsLatest) {
+                    logDebug("Org does not exist. 404 error, exiting...");
+                    setInvalidUrl(true);
+                    setReady(true);
+                    return;
+                }
+
+                if (!org && !viewerIsLatest) {
```

Comment:
@AfeezGL Have you experienced a scenario where the local storage has a stale `viewer`?

Reply:
@AfeezGL That sounds plausible. We should probably combine all this to a single call that accepts `orgSlug`, and an optional `projectId/Slug` then returns a user object with both the current org. and project.

Reply:
@jdamilola Okay, I'll work on that.